### ISSUE TITLE
ambiguous archive policy matching if duplicate patterns

### ISSUE THREAD

**Issue by chungg** [gnocchixyz/gnocchi#199]
we allow archive policy rule pattern to be duplicated. this creates ambiguity as we have no idea which rule it will pick... currently it matches with whatever order sql returns back (i assume create time).

we should not let duplicate archive rules with same pattern  OR we need to allow it to be scope to project (and still not let it have same pattern)

**Comment by jd** [comment]
Wrong, they are ordered by pattern. Which barely make any sense anyway, they should at least by ordered by name. It's pretty common to match based on that ordering.

**Comment by chungg** [comment]
yeah, well that's my point. you can have more than one rule that match '*' pattern... but that doesn't make any sense because multiple rules can't be applied to a single metric... or that's the point of my bug

**Comment by Asu4ni** [comment]
Instead of picking a rule out of all the matched ones, is it reasonable to simply reject the post query with error status due to ambiguity?

**Comment by jd** [comment]
Ordering by name would be enough, with a first match first pick.

I honestly don't like this archive policy rule and I'm hoping we can phase it out. But currently it's needed until we fix issues like #222.

**Comment by Asu4ni** [comment]
After discussing with my colleagues, we think that ordering the rules by the time they are built should be more reasonable. The rule made the latest would be the choice then. The implementation, therefore, should either use the timestamp stored in the database or add a new attribute of time built to each rules.

By the way, I'm wondering if archive policy rule might probably be removed in the future and 'default_archive_policy' in ceilometer.conf will be removed as well, which policy will it use when the metrics sent from ceilometer? Are there any ways to specify the policy for each metrics I'm missing? And what if some metrics have no policy specified?

**Comment by jd** [comment]
Time is unreliable and does not allow to change the order after you created them, so it's a terrible idea.

This is not the right place to discuss Ceilometer thing, but default_archive_policy will not be removed, and ultimately if archive policy rules are phased out, it'd be Ceilometer responsibility to build its own system.

**Comment by Asu4ni** [comment]
I understand that the problem with time. But still, using name is kind of awkward, too. Changing the name of rule just to meet the need is really weird, and some weird rule names would appear.
And as for the ceilometer problems, let us discussed elsewhere! :)

**Comment by chungg** [comment]
i guess we can close?

**Comment by jd** [comment]
TBH @Asu4ni showed us how you can still have problems. I don't think the current solution is good enough. I think we really should have something that has no weird effect like sorting by name only.

Or we go ahead and deprecate this entirely?

**Comment by chungg** [comment]
i'm not sure how sorting by name only solves issue? it will ultimately defer to the pattern regex anyway? we should really be ordering by regex length i think

my original issue was that we for some reason allow user to define two patterns that are exactly the same which i think now is predictable but as you mentioned is not perfect.

**Comment by jd** [comment]
Sorting by names solves the issue because you can control the order of how the regexes are applied. Right now you can't, so there's no way to make the example from @Asu4ni work. With a sorting order you control, you can.

Your original issue is a specialization of the problem, but the general one still applies.
